File: dal1.js
// nodejs_min_max_compare_comparisons.js

const { performance } = require("perf_hooks");

// Brute Force: hitung jumlah komparasi secara eksplisit
function bruteForce(arr) {
  let min = arr[0],
    max = arr[0];
  let comparisons = 0;

  for (let i = 1; i < arr.length; i++) {
    comparisons++;
    if (arr[i] < min) {
      min = arr[i];
    }
    comparisons++;
    if (arr[i] > max) {
      max = arr[i];
    }
  }

  return { min, max, comparisons };
}

// Divide and Conquer: hitung komparasi dalam objek counter
function divideAndConquer(arr, low, high, counter) {
  if (low === high) {
    return { min: arr[low], max: arr[low] };
  }

  if (high === low + 1) {
    counter.count++;
    if (arr[low] < arr[high]) {
      return { min: arr[low], max: arr[high] };
    } else {
      return { min: arr[high], max: arr[low] };
    }
  }

  const mid = Math.floor((low + high) / 2);
  const left = divideAndConquer(arr, low, mid, counter);
  const right = divideAndConquer(arr, mid + 1, high, counter);

  // dua komparasi untuk menggabungkan min dan max
  counter.count++;
  const min = left.min < right.min ? left.min : right.min;

  counter.count++;
  const max = left.max > right.max ? left.max : right.max;

  return { min, max };
}

function dac(arr) {
  const counter = { count: 0 };
  const result = divideAndConquer(arr, 0, arr.length - 1, counter);
  return { ...result, comparisons: counter.count };
}

// Ukuran array yang akan diuji
const sizes = [1000, 10000, 100000, 1000000, 10000000];
const results = [];

for (const N of sizes) {
  // Generate array acak
  const arr = Array.from({ length: N }, () =>
    Math.floor(Math.random() * 10000000)
  );

  const bf = bruteForce(arr);
  const dnq = dac(arr);

  results.push({
    N,
    algo_BF: bf.comparisons,
    algo_DNQ: dnq.comparisons,
  });
}

console.table(results);

File: dal1.py
import random
import matplotlib.pyplot as plt

# Brute Force: hitung jumlah perbandingan
def brute_force_count(arr):
    minimum = arr[0]
    maximum = arr[0]
    comparisons = 0

    for i in range(1, len(arr)):
        # satu perbandingan untuk min
        comparisons += 1
        if arr[i] < minimum:
            minimum = arr[i]
        # satu perbandingan untuk max
        comparisons += 1
        if arr[i] > maximum:
            maximum = arr[i]

    return comparisons

# Divide and Conquer: hitung jumlah perbandingan dan hasil min/max
def dac_count(arr, low, high):
    # kasus dasar satu elemen
    if low == high:
        return arr[low], arr[low], 0

    # kasus dasar dua elemen
    if high == low + 1:
        comparisons = 1
        if arr[low] < arr[high]:
            return arr[low], arr[high], comparisons
        else:
            return arr[high], arr[low], comparisons

    # rekursi
    mid = (low + high) // 2
    lmin, lmax, lcmp = dac_count(arr, low, mid)
    rmin, rmax, rcmp = dac_count(arr, mid + 1, high)

    # gabungkan hasil
    comparisons = lcmp + rcmp
    comparisons += 1  # perbandingan untuk min
    current_min = lmin if lmin < rmin else rmin
    comparisons += 1  # perbandingan untuk max
    current_max = lmax if lmax > rmax else rmax

    return current_min, current_max, comparisons

# Wrapper untuk memanggil D&C
def divide_and_conquer_count(arr):
    _, _, comparisons = dac_count(arr, 0, len(arr) - 1)
    return comparisons

# Uji berbagai ukuran array
sizes = [1000, 10000, 100000, 1000000, 10000000]
bf_counts = []
dc_counts = []

for N in sizes:
    arr = [random.randint(1, 10000000) for _ in range(N)]
    bf_counts.append(brute_force_count(arr))
    dc_counts.append(divide_and_conquer_count(arr))

# Plot hasil
plt.figure()
plt.plot(sizes, bf_counts, label='Brute Force')
plt.plot(sizes, dc_counts, label='Divide and Conquer')
plt.xlabel('Panjang Array (N)')
plt.ylabel('Jumlah Komparasi')
plt.title('Perbandingan Jumlah Komparasi Algoritma BF vs D&C')
plt.legend()
plt.tight_layout()
plt.show()